MongoCursor RxJava observable wrapper

philleonard opened the pull request:
Introduced a simple RxJava Observable wrapper to expose the functionality of RxJava in `MongoCursor`. Also added a default Flowable with a buffer back pressure strategy as it is assumed that data from a query be buffered if the downstream cannot keep up.

Imagine a zip on a slow network call and a quick Jongo call;

```
Flowable.zip(networkFlowable, cursor.toFlowable(), (net, db) -> {
            return combine(net, db); // Some operation
});
```

Back pressure is applied to the consumption of the Mongo query and the items are buffered.

ctrimble commented:
I wonder if this is a place where it might be better to add a method like this to MongoCursor:

```
public <R> as( Function<MongoCursor, R> toType ) {
  return toType.apply(this);
}
```

And then have an optional dependency on RxJava with something like:

```
public class RxJongo {
  Function<MongoCursor, Observable> observable() {
    //wrapper logic here.
  }
}
```

Then you could do:

```
cursor.as(RxJongo.observable())
```

This would give the same effect, but without the direct coupling.

philleonard commented:
This is a nice decoupling solution. Should we wait for @bguerout to drop by and comment?

bguerout commented:
Hello
Thank you for your pull request.
You are right it should be nice to find a decoupled solution but the `as` method with `Function` param required JDK 8.
@ctrimble you suggested that #301 can help. Do you mean consumer interface ?

ctrimble commented:
I don't think in this case you would want a consumer interface.  Since the goal here is to transform the MongoCursor into some other representation, it seems like a method that accepts a transformation function is the right way to go.

I opened #301 to generally see where we could add methods that take SAML interfaces.  This seems like a good spot.

ctrimble commented:
One more point, Jongo can just supply its own `Function<T,R>`, or it could pick up a dependency on something like [Lightweight-Stream-API](https://github.com/aNNiMON/Lightweight-Stream-API), until the project decides to move to Java 8.  RetroLambda could be used to simplify implementation.

philleonard commented:
Thanks for your input guys. I have moved to a decoupled implementation with a `Function<T, R>` interface. I have also implemented it in a way that the observable (and flowable) emission type is preserved.

ctrimble commented on pom.xml:
```diff
@@ -124,6 +125,11 @@
+        <dependency>
+            <groupId>io.reactivex.rxjava2</groupId>
+            <artifactId>rxjava</artifactId>
+            <version>${rx.version}</version>
```
This should probably be optional.

philleonard replied:
Yes, sorry it should.

ctrimble commented on src/main/java/org/jongo/Function.java:
```diff
@@ -0,0 +1,18 @@
+package org.jongo;
```
I suggest we place these interfaces in `org.jongo.function`.

ctrimble commented on src/main/java/org/jongo/RxJongo.java:
```diff
@@ -0,0 +1,46 @@
+package org.jongo;
```
Should this be in a sub package of some kind?

philleonard replied:
Maybe `org.jongo.transformer` or `org.jongo.util`?

ctrimble replied:
I am not sure...  In the past I have kept code with optional dependencies in sub packages, as a `best practice`, due to other tooling commonly found in Java tool chains.  This may no longer be a concern.

ctrimble replied:
Since I cannot remember the exact technical conditions under which placing this in `org.jongo` is bad, it does not seem like that can be a factor.  I would probably look for something like this in the `org.jongo` or `org.jongo.util`, before I looked in `org.jongo.transformer`.  Perhaps it is best to leave it in `org.jongo`, until there are a few of them and they must be moved to a separate package.

philleonard replied:
Ok sure, makes sense.

bguerout replied:
I think the RxJongo implementation should not be packaged in Jongo. 
The new Function interface allows someone to create its own implementation (reactivex or other). 
We can add a section in the documentation to explain how to add reactiveX with the code you have provided.

philleonard replied:
I can understand, my motivation was that this would save others from having to implement the same generic wrapper.

bguerout replied:
Note that there is another repository : https://github.com/bguerout/jongo-extras with other implementations/recipes.
It should be nice to add a RxJongo folder.
What do you think about that ?

philleonard replied:
Ok that looks like a better place for it to live. There appears to be no maven artifact for the extras library though... But I will leave the functional interface in here and then move the Rx implementation to the extras project.

ctrimble commented:
LGTM

philleonard commented:
Is it possible to wrap this one up?
